### PR TITLE
Restructure search engine settings #1945 also fixes #2010 and #1979

### LINKED ISSUE #2010: Dismissing "Add search engine" leaves keyboard visible
STR:
- Settings -> Google default -> Installed search engines -> Add another...
- Click text field to raise keyboard
- Hit X

Expected: keyboard closes
Actual: keyboard stays open

Like #2011.

### PR DESCRIPTION
**Pull request by markushi**
The `SettingsFragment` currently has a lot of responsibilities. Besides the main entry point it also is used for the showing the "Installed search engines", "Add search engine" and "Remove search engines" screens.

This PR splits every screen into separate Fragments (#1945). 
Additionally the keyboard now gets properly shown/hidden for the "Add search engine" screen (#2010 and #1979).

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/settings/SettingsFragment.java [hunk 1]
```diff
@@ -4,173 +4,64 @@
...
 import org.mozilla.focus.widget.DefaultBrowserPreference;
 
 import java.util.Locale;
-import java.util.Set;
 
-public class SettingsFragment extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener {
-    public static final String SETTINGS_SCREEN_NAME = "settingsScreenName";
-    public static final int ALPHA_ENABLED = 255;
-    public static final int ALPHA_DISABLED = 130;
+public class SettingsFragment extends BaseSettingsFragment implements SharedPreferences.OnSharedPreferenceChangeListener {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/settings/SettingsFragment.java]: \[PMD\] INFO: The class 'SettingsFragment' has a Cyclomatic Complexity of 3 \(Highest = 7\)\.

**Comment by pocmo** [comment]
This PR is looking really good!

detekt is complaining about 3 small things:
```
        MaxLineLength - [RemoveSearchEnginesSettingsFragment.kt] at focus-android/app/src/main/java/org/mozilla/focus/settings/RemoveSearchEnginesSettingsFragment.kt:42:167
        MaxLineLength - [RemoveSearchEnginesSettingsFragment.kt] at focus-android/app/src/main/java/org/mozilla/focus/settings/RemoveSearchEnginesSettingsFragment.kt:52:130
        NewLineAtEndOfFile - [BaseSettingsFragment.kt] at focus-android/app/src/main/java/org/mozilla/focus/settings/BaseSettingsFragment.kt:36:1

```

**Comment by markushi** [comment]
Thanks for taking a look!
Yeah my bad, my dev environment wasn't correctly set up. Should be fine now!

File: app/src/main/java/org/mozilla/focus/settings/SettingsFragment.java (hunk 1 above)
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/settings/SettingsFragment.java]: \[PMD\] INFO: The class 'SettingsFragment' has a Cyclomatic Complexity of 3 \(Highest = 7\)\.

File: app/src/main/java/org/mozilla/focus/settings/BaseSettingsFragment.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package org.mozilla.focus.settings
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/settings/BaseSettingsFragment.kt]: nit: Please add our license header:

```Kotlin
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```

File: app/src/main/java/org/mozilla/focus/settings/BaseSettingsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        }
+    }
+
+    protected fun getActionBarUpdater() = activity as ActionBarUpdater
+
+    protected fun getSearchEngineSharedPreferences(): SharedPreferences {
+        return activity.getSharedPreferences(SearchEngineManager.PREF_FILE_SEARCH_ENGINES, Context.MODE_PRIVATE)
+    }
+
+    protected fun pushFragment(fragment: Fragment) {
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/settings/BaseSettingsFragment.kt]: Not sure if this helper method adds more clarity (e.g. if reading pushFragment() in other places). Maybe this just needs a more descriptive name?

File: app/src/main/java/org/mozilla/focus/settings/InstalledSearchEnginesSettingsFragment.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
+package org.mozilla.focus.settings
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/settings/InstalledSearchEnginesSettingsFragment.kt]: nit: license header.

File: app/src/main/java/org/mozilla/focus/settings/RemoveSearchEnginesSettingsFragment.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package org.mozilla.focus.settings
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/settings/RemoveSearchEnginesSettingsFragment.kt]: nit: license header

File: app/src/main/java/org/mozilla/focus/settings/BaseSettingsFragment.kt (hunk 3 above)
Comment (markushi) [reply app/src/main/java/org/mozilla/focus/settings/BaseSettingsFragment.kt]: makes sense, I renamed it to `navigateToFragment()` to make the purpose more clear - what do you think?

**Pull request closed without merging**